lib/dpl/provider/pages.rb

```diff
@@ -44,34 +53,140 @@ def slug
       end
 
+      def keep_history
+        options.fetch(:keep_history) { false }
+      end
+
+      def allow_empty_commit
+        options.fetch(:allow_empty_commit) { false }
+      end
+
+      def verbose
+        # Achtung! Never verbosify git, since it may expose user's token.
+        options.fetch(:verbose) { false }
+      end
+
       def check_auth
       end
 
       def needs_key?
         false
       end
 
-      def github_deploy
-        context.shell 'rm -rf .git > /dev/null 2>&1'
-        context.shell "touch \"deployed at `date` by #{@gh_name}\""
-        context.shell 'git init' or raise 'Could not create new git repo'
+      def print_step(msg)
+        $stderr.puts msg if @verbose
+      end
+
+      def github_pull(target_dir)
+        print_step "Trying to clone a single branch #{@target_branch} from existing repo..."
+        unless context.shell "git clone --quiet --branch='#{@target_branch}' --single-branch '#{@gh_remote_url}' '#{target_dir}' &>/dev/null"
+          # if such branch doesn't exist at remote, do normal clone and create
+          # a new orphan branch
+          print_step "Cloning #{@target_branch} branch failed"
+          print_step 'Trying to clone the whole repo...'
+          context.shell "git clone --quiet '#{@gh_remote_url}' '#{target_dir}' &>/dev/null" or raise "It looks the repo doesn't exist on remote or is inaccessible"
+          FileUtils.cd(target_dir, :verbose => @verbose) do
+            print_step "Assuming #{@target_branch} branch doesn't exist, thus creating orphan one"
+            context.shell "git checkout --orphan '#{@target_branch}'"
+          end
+        end
+      end
+
+      def github_clean
+        context.shell "pwd"
+        print_step 'Purging every existing file from repo...'
+        context.shell "git ls-files -z 2>/dev/null | xargs -0 rm -f 2>/dev/null"  # remove all committed files from the repo
+        print_step 'Cleaning up all folders from repo...'
+        context.shell "git ls-tree --name-only -d -r -z HEAD 2>/dev/null | sort -rz | xargs -0 rmdir 2>/dev/null"  # remove all directories from the repo
+      end
+
+      def github_init
+        print_step 'Creating a brand new local repo from scratch...'
+        context.shell "git init" or raise 'Could not create new git repo'
+        print_step 'Repo created successfully'
+        context.shell "git checkout --orphan '#{@target_branch}'" or raise 'Could not create new git repo'
+        print_step "An orphan branch #{@target_branch} created successfully"
+      end
+
+      def github_configure
+        print_step "Configuring git committer to be #{@gh_name} <#{@gh_email}>"
         context.shell "git config user.email '#{@gh_email}'"
         context.shell "git config user.name '#{@gh_name}'"
+      end
+
+      def github_commit
+        print_step "Preparing to deploy #{@target_branch} branch to gh-pages"
+        context.shell "touch \"deployed at `date` by #{@gh_name}\""
```

Comment:
This line makes `--allow_empty_commit` have no effect, since there will always be a 1-file change. That, the existence of #677, #696, and #697, and the fact that I can't think of a use case for this file that's not satisfied by the commit itself, are why I just removed the line in #715.

Reply:
@jyasskin I think, I saw this in the one of guides I read while researching prior art. Now that I see lots of people are not satisfied with it I think we can safely remove it. On the other hand, I don't want to mix lots of changes in a single PR.